sda_download/sda_download.go

```diff
@@ -0,0 +1,259 @@
+		if err != nil {
+			return err
+		}
+
+		filePathSplit := strings.Split(filePath, "/")
+		if strings.Contains(filePath, "elixir-europe.org") {
+			filePath = strings.Join(filePathSplit[1:], "/")
+		}
+
+		outFilename := filePath
+		if *outDir != "" {
+			outFilename = *outDir + "/" + filePath
+		}
+
+		err = downloadFile(downloadurl, config.AccessToken, outFilename)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// downloadFile downloads the file by using the download URL
+func downloadFile(uri, token, filePath string) error {
+	filePath = strings.TrimSuffix(filePath, ".c4gh")
+	// Get the file body
+	body, err := getResponseBody(uri, token)
+	if err != nil {
+		return fmt.Errorf("failed to get file for download, reason: %v", err)
+	}
+
+	// Create the directory if it does not exist
+	fileDir := filepath.Dir(filePath)
+	err = os.MkdirAll(fileDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create directory, reason: %v", err)
+	}
+
+	outfile, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to create file, reason: %v", err)
+	}
+	defer outfile.Close()
+
+	// Create a new progress container
+	p := mpb.New()
+
+	// Create a new progress bar with the length of the body
+	bar := p.AddBar(int64(len(body)),
+		mpb.PrependDecorators(
+			decor.CountersKibiByte("% .2f / % .2f"),
+		),
+	)
+
+	// Create a proxy reader
+	reader := strings.NewReader(string(body))
+	proxyReader := bar.ProxyReader(reader)
+
+	fmt.Printf("Downloading file to %s\n", filePath)
+	// Copy from the proxy reader (which updates the progress bar) to the file
+	_, err = io.Copy(outfile, proxyReader)
+	if err != nil {
+		return fmt.Errorf("failed to write file, reason: %v", err)
+	}
+
+	// Wait for the progress bar to finish
+	p.Wait()
+
+	return nil
+}
+
+// downloadURL gets the datset files, parses the JSON response to get the file ID
+// and returns the download URL for the file
+func downloadURL(baseURL, token, dataset, filename string) (string, error) {
+	// Sanitize the base_url
+	baseURL = strings.TrimSuffix(baseURL, "/")
+	if !strings.HasPrefix(baseURL, "http") {
+		return "", fmt.Errorf("invalid URL, missing protocol (http/https)")
+	}
+
+	// Make the url for listing files
+	filesURL := baseURL + "/metadata/datasets/" + dataset + "/files"
+
+	// Get the response body from the files API
+	body, err := getResponseBody(filesURL, token)
+	if err != nil {
+		return "", fmt.Errorf("failed to get files, reason: %v", err)
+	}
+
+	// Parse the JSON response
+	var files []File
+	err = json.Unmarshal(body, &files)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse file list JSON, reason: %v", err)
+	}
+
+	// Get the file ID for the filename
+	fileID := ""
+	for _, file := range files {
+		if strings.Contains(file.FilePath, filename) {
+			fileID = file.FileID
+
+			break
+		}
+	}
+
+	if fileID == "" {
+		return "", fmt.Errorf("failed to find file ID for %s", filename)
+	}
```

Comment:
You don't need to iterate through the list to find the entry you want.
something like this should solve it:
```go
idx := slices.IndexFunc(myconfig, func(f files) bool { return f.FileName == filename })
if idx == -1 {
    return "", fmt.Errorf("failed to find file ID for %s", filename)
}

then use `files[idx]` in the return statement at the end.
```

This should be expanded so that the user can supply either a file ID or a file path.

Reply:
That is cool, I had no clue about the IndexFunc.
Done